### ISSUE TITLE
Importing Image and Metadata via XML

### ISSUE THREAD

**Issue by vk-chaertl** [cobwebch/external_import#243]
Hello, and first of all: Thanks for the amazing extension.

I have a question that basically boils down to: I have no idea what I'm doing. Soooo... thanks in advance. :)

We are trying to import data via XML. It basically has this form, though each project can have 0 or more images in the image tag:
```
<projects>
 <project>
  <name>Hans</name>
  <images>
   <image>
    <base64>(base64 string)</base64>
    <copyright>Photographer Paul</copyright>
    <alt>The picture shows Hans riding a horse.</alt>
   </image>
   <image>
    <base64>(base64 string)</base64>
    <copyright>Photographer Petra</copyright>
    <alt>The picture shows a horse riding on Hans.</alt>
   </image>
  </images>
 </project>
 <project>
  <name>Franz</name>
  <images/>
 </project>
</project>
</projects>
```

Now, I attempted this TCA configuration:

```
$GLOBALS['TCA']['tt_address']['columns']['image']['external'] = [
    0 => [
        'field' => 'images',
        'substructureFields' => [
            10 => [
                'field' => 'base64',
                'transformations' => [
                    20 => [
                        'userFunction' => [
                            'class' => \Cobweb\ExternalImport\Transformation\ImageTransformation::class,
                            'method' => 'saveImageFromBase64',
                            'parameters' => [
                                'storage' => '1:steckbriefe',
                                'nameField' => 'name',
                                'defaultExtension' => 'jpg'
                            ]
                        ]
                    ]
                ],

                'children' => [
                    'table' => 'sys_file_reference',
                    'columns' => [
                        'uid_local' => [
                            'field' => 'image'
                        ],
                        'uid_foreign' => [
                            'field' => '__parent.id__'
                        ],
                        'title' => [
                            'field' => 'title'
                        ],
                        'tablenames' => [
                            'value' => 'tt_address'
                        ],
                        'fieldname' => [
                            'value' => 'image'
                        ],
                        'table_local' => [
                            'value' => 'sys_file'
                        ],
                        'alternative' => [
                            'field' => 'alt'
                        ],
                        'copyright' => [
                            'field' => 'copyright'
                        ]
                    ]
                ]
            ]
        ]
    ]
];
```

What I am trying to do, hopefully obviously, is to fill multiple images into some tt_address entries when I import several tt_addresses, and already fill out some of the metadata that I have saved in the XML.

In inspection, I get the following error:

**The "substructureFields" configuration contains invalid properties (transformations, children) for XML-type data. Only 2$transformations, children are allowed (as well as being empty)**

I guess I am nesting my TCA wrong? Or is it really not possible to do this with XML? I could switch to JSON with some effort.

**Comment by fsuter** [comment]
Hi. Well, your TCA is kind of all screwed up ;-)

The `substructureFields` property is generating an error, because it's not supposed to be an indexed array (that's where you have `10` in your code). If you look at the example in the documentation (https://player.vimeo.com/video/312939280?h=975c5bd59f), the array keys are expected to be field names. Those field can either already exist in the TCA or will be created on the fly. They can also be part of the additional fields.

The `transformations` array has to be at the same level as the `substructureFields` array, not inside it.

The same is true for the `children` configuration.

You also need to point to the `<image>` node, because you want each of these and not the `<images>` parent node. So instead of:
`'field' => 'images'`
you need to have:
`'xpath' => 'images/image'`

There's no example for this in the docs, but there's one in extension "externalimport_test". See: https://github.com/cobwebch/externalimport_test/blob/928a1db27614cb7f56c47bccc496d5b19eac1d56/Configuration/TCA/tx_externalimporttest_product.php#L241

I hope this will help you. I won't be able to answer further in the next weeks, as I will be away. Good luck!

A couple more tips:

- use the Preview mode in the BE module. It's really helpful to check if the data is what you expect at any given step
- related to this, get familiar with the step structure: https://docs.typo3.org/p/cobweb/external_import/6.0/en-us/User/Overview/Index.html and note that each property is described in the documentation with a "Scope", i.e. the step to which it applies. This is helpful to understand what happens at what point to your data, and how some properties apply one after the other, while others apply somewhat simultaneously.

**Comment by vk-chaertl** [comment]
I finally managed to solve it, big time to your explanation of my misunderstandings.

Curiously enough, I had to rewrite the ImageTransformation.php to accept the correct field that I intended it to as the URI, so I think I am still not doing it the way  you originally intended it, but... well, it works now!

Thanks a damn lot! And I hope you are well and not in a dire medical situation, given the first link... All the love from me either way. <3
